Add test to check speed guarantees while degraded

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds the test `TestE2eDevNetWithEspressoEspressoDegradedLivenessViaCaffNode` which asserts the rollup integration requirement that we should be able to provide Espresso Confirmations within 10 seconds of the rollup producing a block, as outlined within our documention for [Rollup Integration Requirements](https://eng-wiki.espressosys.com/mainch16.html#x20-10000016).

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209467161000900

Ayiga committed:
Add test to check speed guarantees while degraded

In order to fulfill the requirements for liveness test for general rollup
integrations, we need to ensure that Espresso is able to provide
confirmations for the L2 within 10 seconds of the block being produced.

To that end we actually test that the degraded simulated
espresso-dev-node is able to provide a 10 seconds confirmation within
10 seconds of the transaction being submitted successfully to the L2.

dailinsubjam commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,293 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+// transaction number we submitted to the sequencer.  As a result, we should
+// be able to match the Batch number to the transaction number in transaction
+// order.
+// Sadly, there does not seem to be an easy way to associate the Transaction
+// submitted to the L2 with the Block being returned from Espresso.  However,
+// in this test scenario, we know that the Batch number will correspond to the
+// transaction number we submitted to the sequencer.  As a result, we should
+// be able to match the Batch number to the transaction number in transaction
+// order.
```
Weird, this paragraph is repeated 4 times by accident?

Ayiga replied:
This is a likely a vim-motion fat fingering by me.  My bad!

Ayiga replied:
Fixed in 677d4f8355c4f03ee5a97a80aa6dde0923788b0b.

dailinsubjam commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,293 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+			t.Errorf("Submission %d was not confirmed in an espresso block within 10 seconds of submission:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", i, diff, want)
+		}
+	}
+
+	averageDuration := totalDiff / N
```
If we did `continue` in line 408 (for whatever reason) then this `totalDiff` will not contain all the `N` diffs right?

Ayiga replied:
This is true, this should probably be being divided by `len(espressoReceipts)`

Ayiga replied:
Addressed in 84676563893006ce81a54b1984a37a7aa24b23a2.

dailinsubjam replied:
Thanks for the commit message! I spent some time thinking about how to ensure that every transaction in espressoReceipts is checked and checked exactly once. Based on the current code logic, I believe this is indeed the case. But adding a few comments around that part of the code might be helpful for other's easy reasoning :)

dailinsubjam commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,293 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		streamer.Refresh(streamBlocksCtx, finalizedL1BlockRef, l2BlockRef.Number)
+
+		// Start consuming Batches from the Streamer
+		// We cannot guarantee the order of these batches coming from the
+		// streamer.  However, luckily, we can rely on the fact that the Batch
+		// number will be stored within the Batch itself, and we can use that
```
Please correct me if I'm wrong, my understanding is that batches in the streamer is already ordered, although I agree it's easier and safer to map them up with the batch number.

Ayiga replied:
You know, this is a **VERY** good point.  When I was first starting this test, I was actually attempting to tie the references together based on the `BlockHash` coming from the `Receipt` of the initial transaction submission to the `BlockHash` on the `Batches` themselves.  Surprisingly, this works for the first batch, but **only** the first batch.

Eventually, I realized that I could take advantage of the submission order to correspond to the batch order, specifically because we wait for the receipt confirmation for each transaction submission before we move on to submitting the next transaction.

But the overall structure / strategy I had created hadn't been updated to reflect that this might be better suited as an ordered slice as a result. But we can now take advantage of that, and potentially simplify this test a bit.

dailinsubjam commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,293 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		})(streamBlocksCtx, &wg, streamer)
+	}
+
+	type submission struct {
+		receipt   *geth_types.Receipt
+		created   time.Time
+		submitted time.Time
+		received  time.Time
+	}
```
Nit: This is a good struct and very clear to refer! Although looks like `created` and `received` is not used here? I don't have a strong preference on keeping or removing them though.

Ayiga replied:
It is true that `created` and `received` are not currently referenced.  I wasn't certain what timestamp to reference to effectively convey the goal of the test, so I just added a few.  It could be interesting to see the differences between these time stamps, but they are ultimately not needed for the test, and may lead to confusion.

At the same time, however, the time that we are really meant to check against would technically be the `received` timestamp, I believe, rather than the `submitted` timestamp, which is what I am currently utilizing.

dailinsubjam reviewed (commented):
Overall looks good to me!

Ayiga committed:
Remove duplicated comment

As pointed out by @dailinsubjam, there is a documentation paragraph that is repeated four times
within our documentation.  This is unnecessary and seems to be an editing oversight.

https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2076763386

Remove the excess repeated paragraphs

Ayiga committed:
Fix averageDuration calculation

The `averageDuration` calculation is dividing by `N` instead of how many entries we
actually encountered, as pointed out by @dailinsubjam.  While this should be fine if everything
works, it does lead to potential errors in computation.  It **should** be the case that the number
of receipts received should correspond to the number of received submissions, so we should
be able to leverage that number instead.

https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2076812347

Fix the averageDuration calculation.

Ayiga committed:
Revert test to check timing using block hash

When designing the test it was noticed that we weren't able to
verify beyond the first block hash for the transactions submitted.
Instead of using this information as an indication of a failure, primarily
due to noticing that the Caff Node was able to verify the receipts
coming through correctly.  Instead it was assumed that the approach
itself was incorrect.

I ended up going choosing an approach that had some faulty
assumptions.  Namely that the transactions we were submitting were
the only batches / l2 blocks being produced.  This is incorrect as
some observation against the tests indicate that 10-16 blocks are
created initially on launch before we even submit any transactions.

This change reverts the test to utilize the Block Hash, as that does
seem to be a pretty decent check for tracking a transaction's
progress through the system.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -2,15 +2,27 @@ package environment_test
 
 import (
 	"context"
```
I needed to add this line to make the code compile (inside my IDE):
```
github.com/ethereum/go-ethereum/common
```

Ayiga replied:
Ooof, missed this import.

Fixed in b6eefeb54d3131358fba1aa5502994c3eb0fa47a.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+// More importantly, this **SHOULD** also continue to be the state even when
+// Espresso is in a degraded state.
+//
+// Sadly, there does not seem to be an easy way to associate the Transaction
```
Could this function help to achieve such a goal? https://github.com/EspressoSystems/optimism-espresso-integration/blob/1a2ec597438badb227320b0bcb58a69eba2c768c/op-e2e/e2eutils/geth/wait.go#L64

Just an idea no need to change your test that is working.

Ayiga replied:
Ah this initial comment was made under the idea that the data that was making up the `Batch` wasn't able to be correlated with the `Block`s coming from the `L2`.  I have been able to determine that I was incorrect in this understanding, and this comment should be adjusted to reflect that.

Ayiga replied:
Ah this initial comment was made under the idea that the data that was making up the Batch wasn't able to be correlated with the Blocks coming from the L2. I have been able to determine that I was incorrect in this understanding, and this comment should be adjusted to reflect that.

Ayiga replied:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+			system.RollupConfig.L2ChainID.Uint64(),
+			batcher.NewAdaptL1BlockRefClient(l1Client),
+			espressoClient.NewClient(server.URL),
+			nil, // TODO(AG)
```
Maybe no need to copy the TODO here.

Ayiga replied:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		streamer.Refresh(streamBlocksCtx, finalizedL1BlockRef, l2BlockRef.Number)
+
+		// Start consuming Batches from the Streamer
+		// We cannot guarantee the that we will receive only the batches that
```
```suggestion
		// We cannot guarantee that we will receive only the batches that
```

Ayiga replied:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+				t.Fatalf("Waiting for L2 tx:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+			}
+
+			// We not have a receipt from the L2 Sequencer, indicating that
```
```suggestion
			// We now have a receipt from the L2 Sequencer, indicating that
```

Ayiga replied:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		t.Fatalf("Expected to received at least many batches as submissions:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	// We'll check that our timings meet or exceed the requirements of the test.
```
Using the receipt to compute the time it takes for a tx to be confirmed by the Caff node is a cool trick!

philippecamacho commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		}
+	}
+
+	if totalDenom > 0 {
```
Should not we require that `totalDenom > 0`?

Ayiga replied:
Sure, there's no harm in doing this check. If we got to this point with `totalDenom` being `0`, then we would have already errored.  But again, no harm in adding another check.

Ayiga replied:
Added a check in cddd08d9f6fe2c65754b2cde2f62a76b8d48920b.

philippecamacho reviewed (approved):
LGTM

dailinsubjam commented on espresso/environment/2_espresso_liveness_test.go:
```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+	const N = 10
+	{
+		for i := 0; i < N; i++ {
+			// Create teh transaction
```
```suggestion
			// Create the transaction
```

Ayiga replied:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.

Ayiga committed:
Make comment adjustments

There are a few comments that contain spelling errors or outdated
information.  They are referenced in these discussions:

https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2078509472
https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2078513254
https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2078514802
https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2078517145
https://github.com/EspressoSystems/optimism-espresso-integration/pull/126#discussion_r2078855718

Update and fix comments with spelling errors or outdated
approaches.

Ayiga committed:
Add missing import

Ayiga committed:
Add an additional check on totalDenom

 If the totalDenom we have computed from reviewing our submissions
 doesn't match up to our total number of submissions, that is another
 indicator of an error / failure within the test.

 This adds another check to verify that totalDenom matches our
 expected submission count.

Ayiga committed:
Add streamer.Refresh call

If we don't `Refresh` the `streamer` occasionally with new
references / checkpoints for the finalized L1 and the Safe L2 then
we don't end up making progress with the streamer beyond the
initial few batches.

We add the Refresh call here to ensure that we are progressing with updated references

dailinsubjam commented:
LGTM! I got 
```
 2_espresso_liveness_test.go:408: Failed to find batch for submission 9:
        have:
        	"false"
        want:
        	"true"
```
When running it locally, but I guess this could be solved by things like increasing the timeout?
